pyomo/contrib/pynumero/interfaces/nlp_projections.py

```diff
@@ -0,0 +1,303 @@
+            self._original_to_projected[self._original_idxs] = self._projected_idxs
+
+    def n_primals(self):
+        return len(self._primals_ordering)
+
+    def primals_names(self):
+        return list(self._primals_ordering)
+
+    def nnz_jacobian(self):
+        if self._nnz_jacobian is None:
+            J = self.evaluate_jacobian()
+            self._nnz_jacobian = len(J.data)
+        return self._nnz_jacobian
+
+    def nnz_hessian_lag(self):
+        if self._nnz_hessian_lag is None:
+            H = self.evaluate_hessian_lag()
+            self._nnz_hessian_lag = len(H.data)
+        return self._nnz_hessian_lag
+
+    def _project_primals(self, default, original_primals):
+        projected_x = default*np.ones(self.n_primals(), dtype=np.float64)
+        projected_x[self._projected_idxs] = original_primals[self._original_idxs]
+        return projected_x
+        
+    def primals_lb(self):
+        return self._project_primals(-np.inf, self._original_nlp.primals_lb())
+
+    def primals_ub(self):
+        return self._project_primals(np.inf, self._original_nlp.primals_ub())
+
+    def init_primals(self):
+        # Todo: think about what to do here if an entry is not defined
+        # for now, we default to NaN, but there may be a better way
+        # (e.g., taking a new initial value in the constructor?)
+        return self._project_primals(np.nan, self._original_nlp.init_primals())
+
+    def create_new_vector(self, vector_type):
+        if vector_type == 'primals':
+            return np.zeros(self.n_primals(), dtype=np.float64)
+        return self._original_nlp.create_new_vector(vector_type)
+
+    def set_primals(self, primals):
+        # here, we keep a local copy of the projected primals so
+        # we can give back these values in get_primals. This might
+        # not be the best idea since we can't really support this
+        # same strategy for other methods (e.g., init_primals) where
+        # we now use NaNs to fill in any "missing" entries.
+        np.copyto(self._projected_primals, primals)
+        original_primals = self._original_nlp.get_primals()
+        original_primals[self._original_idxs] = primals[self._projected_idxs]
+        self._original_nlp.set_primals(original_primals)
+
+    def get_primals(self):
+        original_primals = self._original_nlp.get_primals()
+        self._projected_primals[self._projected_idxs] = original_primals[self._original_idxs]
+        return self._projected_primals
+
+    def get_primals_scaling(self):
+        return self._project_primals(np.nan, self._original_nlp.get_primals_scaling())
+
+    def evaluate_grad_objective(self, out=None):
+        original_grad_objective = self._original_nlp.evaluate_grad_objective()
+        if out is None:
+            return self._project_primals(0.0, original_grad_objective)
+        out[self._projected_idxs] = original_grad_objective[self._original_idxs]
```

Comment:
Do you need to set the indicies not in `self._original_idxs` to 0?

Reply:
good catch. I fixed this, but it does make one wonder what the benefit of the "out" parameter is in this case. Let's discuss this with Michael as well.